apm-agent-core/src/test/java/co/elastic/apm/agent/MockTracer.java

```diff
@@ -57,9 +59,27 @@ public static ElasticApmTracer createRealTracer(Reporter reporter) {
      * the configuration.
      */
     public static ElasticApmTracer createRealTracer(Reporter reporter, ConfigurationRegistry config) {
+
+        // use an object pool that does bookkeeping to allow for extra usage checks
+        TestObjectPoolFactory objectPoolFactory = new TestObjectPoolFactory();
+
         return new ElasticApmTracerBuilder()
             .configurationRegistry(config)
             .reporter(reporter)
+            // use testing bookkeeper implementation here so we will
+            .withObjectPoolFactory(objectPoolFactory)
+            .withLifecycleListener(new LifecycleListener() {
+                @Override
+                public void start(ElasticApmTracer tracer) {
+
+                }
+
+                @Override
+                public void stop() {
+                    // checking proper object pool usage using tracer lifecycle events
+                    objectPoolFactory.checkAllPooledObjectsHaveBeenRecycled();
```

Comment:
Call `MockReporter#assertRecycledAfterDecrementingReferences` here? Otherwise, we're calling `checkAllPooledObjectsHaveBeenRecycled` before the objects have been recycled.

Reply:
As in this case `tracer` is not always an instance of `MockTracer`, I'd rather keep it as is, there is now at least a couple of ways to assert for this:
1. by calling `MockReporter#assertRecycledAfterDecrementingReferences` then `TestObjectPoolFactory#checkAllPooledObjectsHaveBeenRecycled`
2. by calling `MockReporter#assertRecycledAfterDecrementingReferences` then `MockTracer#stop()`, which uses tracer lifecycle listener.

The nice thing here is that when we do call things in the wrong order, we make most of the test fail, which makes it quite obvious.

Reply:
But currently, calling `MockTracer#stop()` would fail when not calling `MockReporter#assertRecycledAfterDecrementingReferences`, wouldn't it?

what about this:
```suggestion
                    if (reporter instanceof MockReporter) {
                        ((MockReporter) reporter).assertRecycledAfterDecrementingReferences();
                        objectPoolFactory.checkAllPooledObjectsHaveBeenRecycled();
                    }
```

Reply:
That's definitely a good idea ! Then we just need to make sure `MockTracer#stop()` is properly called in tests, which could also be asserted, maybe with a JVM shutdown hook (but that might be overkill).